%% BioMed_Central_Tex_Template_v1.05
%%                                      %
%  bmc_article.tex            ver: 1.05 %
%                                       %

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <27 January 2006>           %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		       %%
%%									   %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    

% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     

\def\includegraphic{}
\def\includegraphics{}

\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
%Review style settings
\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

% Begin ...
\begin{document}
\begin{bmcformat}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{A simple MapReduce algorithm for pruning phylogenetic megatrees}
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Ensure \and is entered between all but   %%
%% the last two authors. This will be       %%
%% replaced by a comma in the final article %%
%%                                          %%
%% Ensure there are no trailing spaces at   %% 
%% the ends of the lines                    %%     	
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\author{Rutger A Vos\correspondingauthor$^1$%
       \email{Rutger A Vos\correspondingauthor - Rutger.Vos@naturalis.nl}%
      \and
         Aaron Steele\correspondingauthor$^2$%
         \email{Aaron Steele\correspondingauthor - asteele@berkeley.edu}
      }
      

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address{%
    \iid(1)Naturalis Biodiversity Center, Einsteinweg 2, Leiden, the Netherlands\\
    \iid(2)UC Berkeley, Berkeley, USA
}%

\maketitle

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% The Section headings here are those for  %%
%% a Research article submitted to a        %%
%% BMC-Series journal.                      %%  
%%                                          %%
%% If your article is not of this type,     %%
%% then refer to the Instructions for       %%
%% authors on http://www.biomedcentral.com  %%
%% and change the section headings          %%
%% accordingly.                             %%   
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstract}
        % Do not use inserted blank lines (ie \\) until main body of text.
        \paragraph*{Background:} Here we say some interesting stuff about
        the growing body of megatrees, their application in many research
        fields and the need for good tools for pruning these trees down to the
        set of taxa of interest. We then say that MapReduce is a design 
        pattern that till now has not been applied to phylogenetic tree 
        pruning but that we've come up with a way to leverage it.
      
        \paragraph*{Results:} Here we say that we've implemented our algorithm
        in perl and in clojure and that it yields good results.

        \paragraph*{Conclusions:} We conclude that this is a potentially
        valuable contribution, especially in combination with a software
        stack for taxonomic name resolution upstream and taxonomic grafting
        downstream (a la phylomatic).
\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% The Section headings here are those for  %%
%% a Research article submitted to a        %%
%% BMC-Series journal.                      %%  
%%                                          %%
%% If your article is not of this type,     %%
%% then refer to the instructions for       %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and change the section headings          %%
%% accordingly.                             %% 
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}
% Points to make:
% 	- usefulness of phylogenies in a variety of fields
% 	- growing body of megatrees
% 	- barriers to their application 
% 	- popularity of phylomatic
% 	- tree pruning as a phylotastic component
% 	- different implementations of tree pruning in phylotastic
% 	- the MapReduce design pattern
% 	- the Hadoop framework
% 	- our application of MapReduce/Hadoop to tree pruning
Phylogenetic trees find application in a variety of research fields such as
orthology assignment, gene function prediction, systematics, biodiversity 
informatics and evolutionary comparative analysis. Due to this increasingly 
broad application, more and more researchers whose expertise and research 
interests lie outside of phylogenetics find themselves needing to obtain 
reasonable approximations of phylogeny. At the same time, experts on the 
methodology of phylogenetic inference have been constructing larger and 
larger trees ("megatrees") of, among other higher taxa, mammals (ref), 
fishes (ref) and angiosperms (ref) while community projects on classification 
have resulted in tree-like artefacts such as the NCBI taxonomy (ref), the 
skeleton of the Tree of Life Web Project (ref) and the "green genes" tree (ref). 
However, a recent survey of community practices (ref) has indicated that there 
are considerable barriers to re-use of these megatrees by non-experts. Among 
these barriers is the dearth of tools to prune such trees down to the set of 
taxa of interest. The "Phylomatic" tool (ref), which by default operates on the 
APG3 composite estimate of phylogeny, is the main exception to this situation as 
it provides a convenient way for researchers to obtain subtrees (usually of 
plants) that can be applied towards towards their research questions - which 
explain Phylomatic's popularity as observed in the survey (ref). Noting the 
usefulness of Phylomatic, biologists and informaticists have teamed 
up to formalize a design pattern subsequent to which they are developing a 
standards-compliant "stack" of web services that provides more generalized 
functionality of taxonomic name resolution, tree pruning and grafting, branch 
length estimation and data integration: "Phylotastic". Participants in this
collaboration have developed tree pruning functionality using a variety of
approaches, including triple store and RDBMS queries and implementations in
several open source programming toolkits for phyloinformatics (e.g. DendroPy (ref),
Bio::Phylo (ref), ...). This process also resulted in the development of a tree
pruning algorithm that leverages the MapReduce design pattern for massively 
parallel data analysis (ref). MapReduce is a simple pattern originally developed
by Google that consists of - at a minimum - two steps. In the first step, input
data are transformed by "mappers" into meaningful key/value pairs. As these 
mappers operate in a trivially parallel way, the same key may be emitted multiple 
times (possibly with different values). These emitted result are subsequently 
aggregated into a single datum comprised of a key associated with a list of 
all values observed for that key, and this datum is then passed to the next
step, where parallelized "reducers" whittle down each datum (i.e. a single key
and a list of values) to zero or more key/value pairs. In an optional third
step, "combiners" then operate on these key/value pairs to obtain a final result.
In recent years, MapReduce has found application in some areas of "big data
biology", for example in the Genome Analysis Tool Kit (GATK, ref), where it is
applied to parallelized traversal of aligned short reads obtained by high-throughput
shotgun sequencing ("next generation sequencing"). Here we present a tree pruning
algorithm that operates on megatrees, returning a result in reasonable time by
employing MapReduce-based parallelization. We have implemented this algorithm in
two programming languages (Perl and Clojure) and have launched a prototype web
service (ref - location) that allows users to harvest subtrees from several 
published megatrees.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Results and Discussion %%
%%
\section*{Results and Discussion}

  % we can get this from the readme
  \subsection*{The algorithm}
	Assume we have a tree as shown in Figure 1A. The nodes on this tree are 
	labelled with integers (1-6), which have been applied in a post-order 
	traversal. The important point of this is that a child node will always 
	have a label whose value is a lower number than any of its ancestors. Now
	assume that we want to get a tree that retains tips A, C, F and G such
	that we end up with a tree as shown in Figure 1B. Here are the steps to
	solve this using our algorithm.
	\subsubsection*{Map}
		In the first step, the map function is passed one taxon to retain 
		per call. In response to each of these, we return a list of key value 
		pairs, where each key is a node label for one of the nodes between 
		the taxon and the root of the tree, and the value is the taxon. 
		This is what it emits:

		 (for A:)
		 1 => A
		 3 => A
		 4 => A
		 6 => A
		
		 (for C:)
		 2 => C
		 3 => C
		 4 => C
		 6 => C
		
		 (for F:)
		 5 => F
		 6 => F
		
		 (for G:)
		 5 => G
		 6 => G

	\subsubsection*{Combine}
		In the second step, the outputs of the map function are passed into 
		a combiner that preprocesses the output from map such that 
		each key - that perhaps was emitted multiple times by map - is only seen 
		once, with a list of the values associated with it, i.e. like so:

		 1 => [ A ]
		 2 => [ C ]
		 3 => [ A, C ]
		 4 => [ A, C ]
		 5 => [ F, G ]
		 6 => [ A, C, F, G ]

		In this step we then switch the keys and values around, as a first
		step towards identifying and filtering out the unbranched internal node 
		that was created by pruning tip E (because of the parallelization we 
		cannot know whether that is node 3 or 4). In addition, we will also 
		want to prune out the unbranched internal nodes that were created by 
		pruning B and D (being node 1 and 2, respectively). For that we count 
		how many descendants those nodes have. So two things: switch keys and 
		values, count number of descendants. Because keys and values we emit 
		need to be scalars we concatenate the keys with | and the values with , 
		(for example). Here's the result we would then emit:

		 A       => 1,1 % the first integer is the node ID, the second its tip count
		 C       => 2,1
		 A|C     => 3,2
		 A|C     => 4,2
		 F|G     => 5,2
		 A|C|F|G => 6,4

	\subsubsection*{Reduce}

		Out of the key/value pairs emitted by the combine step we firstly filter 
		out "unbranched internal nodes" such as node 1 and 2, which became 
		"unbranched" by the pruning of B and D, respectively. This is trivial, 
		because we will just not emit any key/value pairs where the value has a
		tip count of one (the integer after the comma). For 3 and 4 this is a bit 
		harder, we know they are both on the path to the root for both A and C, but 
		the only way to know which of these is the MRCA is by recourse to our 
		node labelling scheme: because the labels were applied in post-order, 
		descendants have lower label values than ancestors, and so 3 is the MRCA. 
		Having so reduced the number of key/value pairs, we finally emit:

		 A|C     => 3,2
		 F|G     => 5,2
		 A|C|F|G => 6,4

		This final result is the tree represented as a taxon bipartition table 
		with labels for each implied node retained.

  % this describes at a high level the perl code and web service
  \subsection*{Tolomatic}
  - using tolomatic
  - web front-end
  - performance

  % this describes at a high level Aaron's code
  \subsection*{Clojure implementation on Hadoop}

  Clojure is a dynamic programming language that compiles down to bytecode which is executed on the Java Virtual Machine. It can natively access Java frameworks like Apache Hadoop, making it a good candidate for implementing distributed MapReduce algorithms in an extrememly performant way. In addition to Clojure, our implementation rides on Cascalog, a high performance data processing library for querying "Big Data" on Hadoop using clusters or local machines with the interactive Clojure REPL. As input, our implementation takes two files: The phylogenetic tree that has been transformed and labelled in a post-order traversal from node tip to root, and a file containing the node tips from which to prune. The output is the  taxon bipartition table described above.

%%%%%%%%%%%%%%%%%%%%%%
\section*{Conclusions}
  Further work, integration with phylotastic.
  
%%%%%%%%%%%%%%%%%%
\section*{Methods}

  % implementation details, design choices, e.g. for the input format
  \subsection*{Prototype implementation}
	This subsection describes the initial implementation in Perl,
	the design choices and limitations.

  % implementation details
  \subsection*{Re-implementations}
    This subsection describes the subsequent re-implementations:
    refinement of the Perl prototype, development of web front-end,
    re-implementation in Clojure.
    
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Authors contributions}
    Text for this section \ldots

%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}
  Text for this section \ldots

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%              
%%  Bmc_article.bst  will be used to                       %%
%%  create a .BBL file for submission, which includes      %%
%%  XML structured for BMC.                                %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %% 
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %% 
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{bmc_article} }     % Bibliography file (usually '*.bib' ) 

%%%%%%%%%%%

\ifthenelse{\boolean{publ}}{\end{multicols}}{}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
  \subsection*{Figure 1 - Example tree}
      Figure of a tree before (A) and after (B) pruning

  \subsection*{Figure 2 - Algorithm flowchart}
      Figure of the steps in the algorithm

\end{bmcformat}
\end{document}







